controllers/flux/templates/wait.sh

```diff
@@ -0,0 +1,81 @@
+#!/bin/sh
+
+# This waiting script is run continuously and only updates
+# hosts and runs the job (breaking from the while) when 
+# update_hosts.sh has been populated. This means the pod usually
+# needs to be updated with the config map that has ips!
+
+# We determine the update_hosts.sh is ready when it has content
+count_lines() {
+	lines=$(cat /flux_operator/update_hosts.sh | wc -l)
+	echo $lines
+}
+
+while [ $(count_lines) -lt 2 ];
+do
+    echo "Host updating script not available yet, waiting..."
+    sleep 5s
+done             
+
+# Run to discover hosts
+/bin/sh /flux_operator/update_hosts.sh
+
+# Show host updates
+cat /etc/hosts
+
+# Main host <name>-0
+mainHost="%s"
+
+printf "\n👋 Hello, I'm $(hostname)\n"
+printf "The main host is ${mainHost}\n"
+
+# These actions need to happen on all hosts
+# Configure resources
+mkdir -p /etc/flux/system
+
+# --cores=IDS Assign cores with IDS to each rank in R, so we  assign 1-N to 0
+flux R encode --hosts=%s --cores=%s > /etc/flux/system/R
+printf "\n📦 Resources\n"
+cat /etc/flux/system/R
+
+mkdir -p /etc/flux/imp/conf.d/
+cat <<EOT >> /etc/flux/imp/conf.d/imp.toml
+[exec]
+allowed-users = [ "flux", "root" ]
+allowed-shells = [ "/usr/libexec/flux/flux-shell" ]	
+EOT
+
+printf "\n🦊 Independent Minister of Privilege\n"
+cat /etc/flux/imp/conf.d/imp.toml
+
+# Add a flux user (required)
+useradd flux || printf "flux user already exists\n"
+
+# Generate curve certificate (only need one shared)
+if [ $(hostname) == "${mainHost}" ]; then
+    printf "\n✨ Curve certificate being generated by $(hostname)\n"
+    sudo -u flux flux keygen /mnt/curve/curve.cert
+    cat /mnt/curve/curve.cert
+
+    # Each node needs same munge key
+    cp /etc/munge/munge.key /mnt/curve/munge.key
+
+    # Start flux with the original entrypoint
+    printf "\n🌀 flux start -o --config /etc/flux/config $@\n"
+    flux start -o --config /etc/flux/config $@
+    # Can we show some debug output (perhaps we can have a variable?)
+    # How can we be sure it ran on all nodes?
+else
+
+    # Wait for main node to copy over its key
+    while [ ! -f /mnt/curve/munge.key ];
+    do
+        printf "Shared munge key not available yet, waiting...\n"
+        sleep 5s
+    done
+    cp /mnt/curve/munge.key /etc/munge/munge.key
+    printf "\n🐸 flux start -o --config /etc/flux/config $@\n"
+    flux start -o --config /etc/flux/config $@
```

Comment:
> I assume the $@ argument is the MPI job you mentioned? And it sounds like that actually worked (great!)

Affirmative! The way an operator works, we have a yaml called a custom resource definition (CRD) where a user can put things like the command to run, working directory, and size. So the command from that CRD is passed through to be run there.

For all of the below, note that I removed `--cores` because @trws reports the default based on hwloc is reasonable.

> When run that way, the Flux instance will shut down when the initial program completes. I'm not sure if that is intentional, but if the idea is to start Flux as a service, then you'll probably want to disable the initial program with -o -Ssbroker.rc2_none.

I think that's exactly what we want - for the Kubernetes operator we would want the job to run, and then since it's a batch.Job object (in Kubernetes) when a pod finishes it gets status "Complete" and then when all the pods are complete it will get cleaned up. These are "MiniClusters" (the name for the CRD) created on demand.

Oh wow so a lot of these tweaks seemed to help a ton! I was finally able to run the debug commands - here is output from my new debugging mode:

<details>

<summary>Debugging Mode</summary>

```console
🐸 sudo -u flux flux start -o --config /etc/flux/config -Scron.directory=/etc/flux/system/cron.d   -Stbon.fanout=256   -Srundir=/run/flux   -Sstatedir=/var/lib/flux   -Slocal-uri=local:///run/flux/local   -Slog-stderr-level=6   -Slog-stderr-mode=local   -Sbroker.quorum=0   -Sbroker.quorum-timeout=none   -Sbroker.exit-norestart=42 flux overlay status
broker.info[0]: start: none->join 0.0150761s
broker.info[0]: parent-none: join->init 3.7418e-05s
broker.info[0]: online:  (ranks )
broker.info[0]: overlay auth flux-sample-0 OK
broker.info[0]: accepting connection from flux-sample-2 (rank 2) uuid 8bd29109-f444-4071-ad7f-73721ddf0d71 status full
broker.info[0]: overlay auth flux-sample-0 OK
broker.info[0]: accepting connection from flux-sample-4 (rank 4) uuid 6a8c3179-d9c1-499a-99c4-958c5c4ed018 status full
resource.err[0]: verify: rank 0 (flux-sample-0) has extra resources: core[1-3]
cron.info[0]: synchronizing cron tasks to event heartbeat.pulse
job-manager.info[0]: restart: 0 jobs
job-manager.info[0]: restart: 0 running jobs
job-manager.info[0]: restart: no checkpoint object
broker.info[0]: rc1.0: running /etc/flux/rc1.d/01-sched-fluxion-resource-start
sched-fluxion-resource.warning[0]: create_reader: allowlist unsupported
sched-fluxion-resource.info[0]: populate_resource_db: loaded resources from core's resource.acquire
broker.info[0]: rc1.0: running /etc/flux/rc1.d/02-cron
broker.info[0]: rc1.0: running /etc/flux/rc1.d/02-sched-fluxion-qmanager-start
broker.info[0]: rc1.0: /etc/flux/rc1 Exited (rc=0) 3.5s
broker.info[0]: rc1-success: init->quorum 3.45743s
broker.info[0]: online: flux-sample-0 (ranks 0)
broker.info[0]: quorum-full: quorum->run 0.100283s
0 flux-sample-0: partial
├─ 1 flux-sample-1: offline
├─ 2 flux-sample-2: full
├─ 3 flux-sample-3: offline
├─ 4 flux-sample-4: full
└─ 5 flux-sample-5: offline
broker.info[0]: rc2.0: flux overlay status Exited (rc=0) 0.0s
broker.info[0]: rc2-success: run->cleanup 0.0108852s
broker.info[0]: cleanup.0: flux queue stop --quiet Exited (rc=0) 0.0s
broker.info[0]: cleanup.1: flux job cancelall --user=all --quiet -f --states RUN Exited (rc=0) 0.0s
broker.info[0]: cleanup.2: flux queue idle --quiet Exited (rc=0) 0.0s
broker.info[0]: cleanup-success: cleanup->shutdown 0.0314425s
broker.info[0]: children-complete: shutdown->finalize 0.0283201s
broker.info[0]: rc3.0: running /etc/flux/rc3.d/01-sched-fluxion-qmanager-stop
broker.info[0]: rc3.0: running /etc/flux/rc3.d/02-sched-fluxion-resource-stop
broker.info[0]: rc3.0: /etc/flux/rc3 Exited (rc=0) 0.1s
broker.info[0]: rc3-success: finalize->goodbye 0.146687s
broker.info[0]: goodbye: goodbye->exit 2.8693e-05s

🐸 sudo -u flux flux start -o --config /etc/flux/config -Scron.directory=/etc/flux/system/cron.d   -Stbon.fanout=256   -Srundir=/run/flux   -Sstatedir=/var/lib/flux   -Slocal-uri=local:///run/flux/local   -Slog-stderr-level=6   -Slog-stderr-mode=local   -Sbroker.quorum=0   -Sbroker.quorum-timeout=none   -Sbroker.exit-norestart=42 flux lsattr -v
broker.info[0]: start: none->join 0.00460096s
broker.info[0]: parent-none: join->init 1.8356e-05s
broker.info[0]: online:  (ranks )
kvs.info[0]: restored KVS from checkpoint on 2022-09-12T17:39:19Z
broker.info[0]: overlay auth flux-sample-0 OK
broker.info[0]: accepting connection from flux-sample-2 (rank 2) uuid 220ae17a-3914-4a1d-8309-1164cf688ae7 status full
broker.info[0]: overlay auth flux-sample-0 OK
broker.info[0]: accepting connection from flux-sample-4 (rank 4) uuid eadeac5b-58b1-4af3-a8e0-a4c52d67596d status full
resource.err[0]: verify: rank 0 (flux-sample-0) has extra resources: core[1-3]
cron.info[0]: synchronizing cron tasks to event heartbeat.pulse
job-manager.info[0]: restart: 0 jobs
job-manager.info[0]: restart: 0 running jobs
broker.info[0]: rc1.0: running /etc/flux/rc1.d/01-sched-fluxion-resource-start
sched-fluxion-resource.warning[0]: create_reader: allowlist unsupported
sched-fluxion-resource.info[0]: populate_resource_db: loaded resources from core's resource.acquire
broker.info[0]: rc1.0: running /etc/flux/rc1.d/02-cron
broker.info[0]: rc1.0: running /etc/flux/rc1.d/02-sched-fluxion-qmanager-start
broker.info[0]: rc1.0: /etc/flux/rc1 Exited (rc=0) 0.4s
broker.info[0]: rc1-success: init->quorum 0.387865s
broker.info[0]: online: flux-sample-0 (ranks 0)
broker.info[0]: quorum-full: quorum->run 0.111895s
broker.exit-norestart                   42
broker.mapping                          (vector,(0,6,1))
broker.pid                              133
broker.quorum                           0
broker.quorum-timeout                   none
broker.rc1_path                         /etc/flux/rc1
broker.rc3_path                         /etc/flux/rc3
broker.starttime                        1663004359.80
conf.connector_path                     /usr/lib/flux/connectors
conf.exec_path                          /usr/libexec/flux/cmd
conf.module_path                        /usr/lib/flux/modules
conf.pmi_library_path                   /usr/lib/flux/libpmi.so
conf.shell_initrc                       /etc/flux/shell/initrc.lua
conf.shell_pluginpath                   /usr/lib/flux/shell/plugins
config.path                             /etc/flux/config
content.backing-module                  content-sqlite
content.blob-size-limit                 1073741824
content.flush-batch-limit               256
content.hash                            sha1
content.purge-old-entry                 10
content.purge-target-size               16777216
cron.directory                          /etc/flux/system/cron.d
hostlist                                flux-sample-[0-5]
instance-level                          0
local-uri                               local:///run/flux/local
log-count                               63
log-critical-level                      2
log-filename                            -
log-forward-level                       7
log-level                               7
log-ring-size                           1024
log-ring-used                           63
log-stderr-level                        6
log-stderr-mode                         local
parent-kvs-namespace                    -
parent-uri                              -
rank                                    0
rundir                                  /run/flux
security.owner                          1234
size                                    6
statedir                                /var/lib/flux
tbon.descendants                        5
tbon.endpoint                           tcp://flux-sample-0:8050
tbon.fanout                             256
tbon.level                              0
tbon.maxlevel                           1
tbon.parent-endpoint                    -
tbon.prefertcp                          0
tbon.tcp_user_timeout                   0s
tbon.torpid_max                         30s
tbon.torpid_min                         5s
tbon.zmqdebug                           0
version                                 0.38.0-29-gb13e64ddd
broker.info[0]: rc2.0: flux lsattr -v Exited (rc=0) 0.0s
broker.info[0]: rc2-success: run->cleanup 0.0133762s
broker.info[0]: cleanup.0: flux queue stop --quiet Exited (rc=0) 0.0s
broker.info[0]: cleanup.1: flux job cancelall --user=all --quiet -f --states RUN Exited (rc=0) 0.0s
broker.info[0]: cleanup.2: flux queue idle --quiet Exited (rc=0) 0.0s
broker.info[0]: cleanup-success: cleanup->shutdown 0.0261153s
broker.info[0]: children-complete: shutdown->finalize 0.0793791s
broker.info[0]: rc3.0: running /etc/flux/rc3.d/01-sched-fluxion-qmanager-stop
broker.info[0]: rc3.0: running /etc/flux/rc3.d/02-sched-fluxion-resource-stop
broker.info[0]: rc3.0: /etc/flux/rc3 Exited (rc=0) 0.1s
broker.info[0]: rc3-success: finalize->goodbye 0.0974202s
broker.info[0]: goodbye: goodbye->exit 2.0318e-05s

🐸 sudo -u flux flux start -o --config /etc/flux/config -Scron.directory=/etc/flux/system/cron.d   -Stbon.fanout=256   -Srundir=/run/flux   -Sstatedir=/var/lib/flux   -Slocal-uri=local:///run/flux/local   -Slog-stderr-level=6   -Slog-stderr-mode=local   -Sbroker.quorum=0   -Sbroker.quorum-timeout=none   -Sbroker.exit-norestart=42 flux dmesg
broker.info[0]: start: none->join 0.00439311s
broker.info[0]: parent-none: join->init 2.0994e-05s
broker.info[0]: online:  (ranks )
broker.info[0]: overlay auth flux-sample-0 OK
broker.info[0]: accepting connection from flux-sample-4 (rank 4) uuid 666b9286-8ed5-4724-b092-84438c2842ca status full
kvs.info[0]: restored KVS from checkpoint on 2022-09-12T17:39:20Z
broker.info[0]: overlay auth flux-sample-0 OK
broker.info[0]: accepting connection from flux-sample-2 (rank 2) uuid d30ccf50-ba14-47ac-967b-779f0c976859 status full
resource.err[0]: verify: rank 0 (flux-sample-0) has extra resources: core[1-3]
cron.info[0]: synchronizing cron tasks to event heartbeat.pulse
job-manager.info[0]: restart: 0 jobs
job-manager.info[0]: restart: 0 running jobs
broker.info[0]: rc1.0: running /etc/flux/rc1.d/01-sched-fluxion-resource-start
sched-fluxion-resource.warning[0]: create_reader: allowlist unsupported
sched-fluxion-resource.info[0]: populate_resource_db: loaded resources from core's resource.acquire
broker.info[0]: rc1.0: running /etc/flux/rc1.d/02-cron
broker.info[0]: rc1.0: running /etc/flux/rc1.d/02-sched-fluxion-qmanager-start
broker.info[0]: rc1.0: /etc/flux/rc1 Exited (rc=0) 0.6s
broker.info[0]: rc1-success: init->quorum 0.578325s
broker.info[0]: online: flux-sample-0 (ranks 0)
broker.info[0]: quorum-full: quorum->run 0.100228s
2022-09-12T17:39:20.537834Z broker.debug[0]: insmod connector-local
2022-09-12T17:39:20.537891Z broker.info[0]: start: none->join 0.00439311s
2022-09-12T17:39:20.537925Z broker.info[0]: parent-none: join->init 2.0994e-05s
2022-09-12T17:39:20.538686Z broker.info[0]: online:  (ranks )
2022-09-12T17:39:20.538656Z connector-local.debug[0]: allow-guest-user=true
2022-09-12T17:39:20.538664Z connector-local.debug[0]: allow-root-owner=true
2022-09-12T17:39:20.556332Z broker.debug[0]: insmod barrier
2022-09-12T17:39:20.556491Z broker.info[0]: overlay auth flux-sample-0 OK
2022-09-12T17:39:20.556798Z broker.info[0]: accepting connection from flux-sample-4 (rank 4) uuid 666b9286-8ed5-4724-b092-84438c2842ca status full
2022-09-12T17:39:20.565441Z broker.debug[0]: insmod content-sqlite
2022-09-12T17:39:20.566122Z broker.debug[0]: content backing store: enabled content-sqlite
2022-09-12T17:39:20.577509Z broker.debug[0]: insmod kvs
2022-09-12T17:39:20.578037Z kvs.info[0]: restored KVS from checkpoint on 2022-09-12T17:39:20Z
2022-09-12T17:39:20.586914Z broker.debug[0]: insmod kvs-watch
2022-09-12T17:39:20.600968Z broker.debug[0]: insmod resource
2022-09-12T17:39:20.611898Z broker.info[0]: overlay auth flux-sample-0 OK
2022-09-12T17:39:20.612835Z broker.info[0]: accepting connection from flux-sample-2 (rank 2) uuid d30ccf50-ba14-47ac-967b-779f0c976859 status full
2022-09-12T17:39:20.685756Z resource.err[0]: verify: rank 0 (flux-sample-0) has extra resources: core[1-3]
2022-09-12T17:39:20.686292Z resource.debug[0]: reslog_cb: resource-init event posted
2022-09-12T17:39:20.686666Z resource.debug[0]: reslog_cb: resource-define event posted
2022-09-12T17:39:20.693746Z broker.debug[0]: insmod cron
2022-09-12T17:39:20.694081Z cron.info[0]: synchronizing cron tasks to event heartbeat.pulse
2022-09-12T17:39:20.703556Z broker.debug[0]: insmod job-manager
2022-09-12T17:39:20.704119Z job-manager.info[0]: restart: 0 jobs
2022-09-12T17:39:20.704128Z job-manager.info[0]: restart: 0 running jobs
2022-09-12T17:39:20.704370Z job-manager.debug[0]: restart: max_jobid=0
2022-09-12T17:39:20.712759Z broker.debug[0]: insmod job-info
2022-09-12T17:39:20.734413Z job-manager.debug[0]: duration-validator: updated expiration to 0.00
2022-09-12T17:39:20.769466Z broker.debug[0]: insmod job-list
2022-09-12T17:39:20.770489Z job-list.debug[0]: job_state_init_from_kvs: read 0 jobs
2022-09-12T17:39:20.790492Z broker.debug[0]: insmod job-ingest
2022-09-12T17:39:20.790720Z job-ingest.debug[0]: configuring with plugins=(null), args=(null)
2022-09-12T17:39:20.790993Z job-ingest.debug[0]: fluid ts=1ms
2022-09-12T17:39:20.800252Z broker.debug[0]: insmod job-exec
2022-09-12T17:39:20.800577Z job-exec.debug[0]: using default shell path /usr/libexec/flux/flux-shell
2022-09-12T17:39:20.800587Z job-exec.debug[0]: using imp path /usr/libexec/flux/flux-imp
2022-09-12T17:39:20.810016Z broker.debug[0]: insmod heartbeat
2022-09-12T17:39:20.811265Z broker.info[0]: rc1.0: running /etc/flux/rc1.d/01-sched-fluxion-resource-start
2022-09-12T17:39:20.910873Z broker.debug[0]: insmod sched-fluxion-resource
2022-09-12T17:39:20.911238Z sched-fluxion-resource.debug[0]: mod_main: resource module starting
2022-09-12T17:39:20.911485Z sched-fluxion-resource.warning[0]: create_reader: allowlist unsupported
2022-09-12T17:39:20.911789Z sched-fluxion-resource.debug[0]: resource graph datastore loaded with rv1exec reader
2022-09-12T17:39:20.911808Z sched-fluxion-resource.info[0]: populate_resource_db: loaded resources from core's resource.acquire
2022-09-12T17:39:20.911974Z sched-fluxion-resource.debug[0]: resource status changed (rankset=[all] status=DOWN)
2022-09-12T17:39:20.911983Z sched-fluxion-resource.debug[0]: mod_main: resource graph database loaded
2022-09-12T17:39:20.912172Z broker.info[0]: rc1.0: running /etc/flux/rc1.d/02-cron
2022-09-12T17:39:20.919589Z broker.info[0]: rc1.0: running /etc/flux/rc1.d/02-sched-fluxion-qmanager-start
2022-09-12T17:39:20.951565Z broker.debug[0]: insmod sched-fluxion-qmanager
2022-09-12T17:39:20.961140Z sched-fluxion-qmanager.debug[0]: service_register
2022-09-12T17:39:20.961167Z sched-fluxion-qmanager.debug[0]: enforced policy (queue=default): fcfs
2022-09-12T17:39:20.961173Z sched-fluxion-qmanager.debug[0]: effective queue params (queue=default): default
2022-09-12T17:39:20.961176Z sched-fluxion-qmanager.debug[0]: effective policy params (queue=default): default
2022-09-12T17:39:20.969291Z sched-fluxion-qmanager.debug[0]: handshaking with sched-fluxion-resource completed
2022-09-12T17:39:20.969377Z job-manager.debug[0]: scheduler: hello
2022-09-12T17:39:20.969500Z job-manager.debug[0]: scheduler: ready unlimited
2022-09-12T17:39:20.969569Z sched-fluxion-qmanager.debug[0]: handshaking with job-manager completed
2022-09-12T17:39:21.116197Z broker.info[0]: rc1.0: /etc/flux/rc1 Exited (rc=0) 0.6s
2022-09-12T17:39:21.116257Z broker.info[0]: rc1-success: init->quorum 0.578325s
2022-09-12T17:39:21.216323Z broker.debug[0]: groups: broker.online=0
2022-09-12T17:39:21.216437Z broker.info[0]: online: flux-sample-0 (ranks 0)
2022-09-12T17:39:21.216496Z broker.info[0]: quorum-full: quorum->run 0.100228s
2022-09-12T17:39:21.218430Z resource.debug[0]: reslog_cb: online event posted
2022-09-12T17:39:21.218578Z sched-fluxion-resource.debug[0]: resource status changed (rankset=[0] status=UP)
broker.info[0]: rc2.0: flux dmesg Exited (rc=0) 0.0s
broker.info[0]: rc2-success: run->cleanup 0.00550471s
broker.info[0]: cleanup.0: flux queue stop --quiet Exited (rc=0) 0.0s
broker.info[0]: cleanup.1: flux job cancelall --user=all --quiet -f --states RUN Exited (rc=0) 0.0s
broker.info[0]: cleanup.2: flux queue idle --quiet Exited (rc=0) 0.0s
broker.info[0]: cleanup-success: cleanup->shutdown 0.030039s
broker.info[0]: online: flux-sample-[0,2,4] (ranks 0,2,4)
broker.info[0]: online: flux-sample-[0,4] (ranks 0,4)
broker.info[0]: children-complete: shutdown->finalize 0.286455s
broker.info[0]: rc3.0: running /etc/flux/rc3.d/01-sched-fluxion-qmanager-stop
broker.info[0]: rc3.0: running /etc/flux/rc3.d/02-sched-fluxion-resource-stop
broker.info[0]: online: flux-sample-0 (ranks 0)
broker.info[0]: rc3.0: /etc/flux/rc3 Exited (rc=0) 0.3s
broker.info[0]: rc3-success: finalize->goodbye 0.260922s
broker.info[0]: goodbye: goodbye->exit 2.337e-05s

💤 sleep infinity
```
</details>

And the flux job runs!

![image](https://user-images.githubusercontent.com/814322/189742476-0e7419fb-93e3-414e-ad8e-5e19441870f9.png)
![image](https://user-images.githubusercontent.com/814322/189742505-908de3fe-902a-4520-be68-540e3202796a.png)

 I'll put in a PR soon and ping you so you can take a look. Thanks again!